New TensorboardLogger, batch2device fix, FilteringCollateFn, minor refactorings.

hexfaker opened the pull request.

Scitator commented on data/collate_fn.py:
```diff
@@ -48,3 +51,26 @@ def stack_non_text_features(data, key):
 
     return data
+
+class FilteringCollateFn():
```
brackets?

hexfaker replied:
Done

Scitator commented on dl/callbacks.py:
```diff
@@ -165,6 +167,56 @@ def on_loader_end(self, state):
+    def __init__(self,
+            logdir,
+            metric_names: List[str] = None,
+            log_on_batch_end=False
```
why don't you add 2 flags? for `batch_end` and `epoch_end`?

hexfaker replied:
Looks like my solution is more flexible. You can log some metrics on batches, some on epochs, and some on both while using two instances of this callback. Supporting same functionality in one instance would introduce unneeded complexity to code of this callback. Now it's simple and quite readable, I'd prefer to keep it that way.

Scitator replied:
Yup, `You can log some metrics on batches, some on epochs` using only **one** instance of callback. I think, it is much easier to think from metrics perspective. So, log some metric only on batch, some only on epoch, another one – both.

hexfaker replied:
@Scitator Same functionality with more code is not good.

Scitator replied:
Agree, thats why I don't like **two** instances approach.

Scitator commented on dl/state.py:
```diff
@@ -9,16 +9,22 @@ class RunnerState(FrozenClass):
     """
 
-    def __init__(self, **kwargs):
+    def __init__(self,
```
self to next below pls

hexfaker replied:
done

hexfaker committed:
Minor refactorings

hexfaker committed:
FilteringCollateFn and batch2device support for non-tensor types

- FilteringCollateFn supports key-value batches and prevents values on
specified keys from being converted to tensor and stacked

- batch2device now skips non-tensor typed items (e.g. stings, list, ndarrays)

hexfaker committed:
GroupTensorboardLogger added
